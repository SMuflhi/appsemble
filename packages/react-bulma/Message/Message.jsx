import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { is } from '../utils';


export default class Message extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    color: PropTypes.string,
    component: PropTypes.string,
  };

  static defaultProps = {
    color: null,
    className: null,
    component: 'article',
  };

  render() {
    const {
      className,
      color,
      component: Component,
      ...props
    } = this.props;

    return (
      <Component
        className={classNames(
          'message',
          is(color),
          className,
        )}
        {...props}
      />
    );
  }
}
