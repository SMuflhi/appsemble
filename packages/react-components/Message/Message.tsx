import { BulmaColor } from '@appsemble/types';
import classNames from 'classnames';
import * as React from 'react';

interface MessageProps {
  /**
   * The message content.
   */
  children: React.ReactNode;

  /**
   * Additional class names to assign to the message element.
   */
  className?: string;

  /**
   * The message type.
   */
  color?: BulmaColor;
}

export default function Message({ children, className, color }: MessageProps): React.ReactElement {
  return (
    <div className={classNames('message', className, { [`is-${color}`]: color })}>
      <div className="message-body">{children}</div>
    </div>
  );
}
