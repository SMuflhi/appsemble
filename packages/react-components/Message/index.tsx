import { type BulmaColor } from '@appsemble/types';
import classNames from 'classnames';
import { type ReactNode } from 'react';

interface MessageProps {
  /**
   * The message content.
   */
  readonly children: ReactNode;

  /**
   * Additional class names to assign to the message element.
   */
  readonly className?: string;

  /**
   * The message type.
   */
  readonly color?: BulmaColor;

  /**
   * An optional header for the message.
   */
  readonly header?: ReactNode;
}

export function Message({ children, className, color, header }: MessageProps): ReactNode {
  return (
    <div className={classNames('message', className, { [`is-${color}`]: color })}>
      {header ? <h6 className="message-header">{header}</h6> : null}
      <div className="message-body">{children}</div>
    </div>
  );
}
