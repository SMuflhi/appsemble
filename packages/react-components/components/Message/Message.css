.root {
  bottom: 10px;
  box-sizing: border-box;
  margin-left: 10px;
  margin-right: 10px;
  position: fixed;
  width: calc(100vw - 20px);
  z-index: 1000;
}

.rootMessage {
  opacity: 1;
  transition: opacity 0.3s ease-in-out;

  &.hidden {
    opacity: 0;
  }
}

.content {
  min-height: 64px;
}
