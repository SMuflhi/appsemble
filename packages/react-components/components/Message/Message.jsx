import { Message as BulmaMessage, MessageBody } from '@appsemble/react-bulma';
import PropTypes from 'prop-types';
import React from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';

import styles from './Message.css';

export default class Message extends React.Component {
  static propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape()),
    remove: PropTypes.func.isRequired,
  };

  static defaultProps = {
    messages: [],
  };

  componentDidUpdate(prevProps) {
    const { messages, remove } = this.props;

    if (messages.length > prevProps.messages.length) {
      const message = messages[messages.length - 1];

      setTimeout(() => {
        remove(message);
      }, 5e3);
    }
  }

  render() {
    const { messages } = this.props;

    return (
      <ul className={styles.root}>
        <ReactCSSTransitionGroup
          transitionEnterTimeout={300}
          transitionLeaveTimeout={300}
          transitionName={{
            enter: styles.messageEnter,
            enterActive: styles.messageEnterActive,
            leave: styles.messageLeave,
            leaveActive: styles.messageLeaveActive,
          }}
        >
          {messages.map(message => (
            <BulmaMessage key={message.id} color={(message && message.color) || 'danger'}>
              <MessageBody className={styles.content}>{message?.body}</MessageBody>
            </BulmaMessage>
          ))}
        </ReactCSSTransitionGroup>
      </ul>
    );
  }
}
