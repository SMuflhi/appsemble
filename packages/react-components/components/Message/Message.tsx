import classNames from 'classnames';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import styles from './Message.css';

interface IncomingMessage {
  body: string;
  color?: 'danger' | 'dark' | 'info' | 'link' | 'primary' | 'success' | 'warning';
  id: number;
}

export interface MessageProps {
  messages: IncomingMessage[];
  remove: (message: IncomingMessage) => void;
}

export default class Message extends React.Component<MessageProps> {
  static propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({})),
    remove: PropTypes.func.isRequired,
  };

  static defaultProps: Partial<MessageProps> = {
    messages: [],
  };

  componentDidUpdate(prevProps: MessageProps): void {
    const { messages, remove } = this.props;

    if (messages.length > prevProps.messages.length) {
      const message = messages[messages.length - 1];

      setTimeout(() => {
        remove(message);
      }, 5e3);
    }
  }

  render(): React.ReactNode {
    const { messages } = this.props;

    return (
      <div className={styles.root}>
        <TransitionGroup>
          {messages.map(message => (
            <CSSTransition
              key={message.id}
              classNames={{
                enter: styles.messageEnter,
                enterActive: styles.messageEnterActive,
                exit: styles.messageExit,
                exitActive: styles.messageExitActive,
              }}
              timeout={300}
            >
              <article
                className={classNames(
                  'message',
                  message && message.color ? `is-${message.color}` : 'is-danger',
                )}
              >
                <div className={classNames('message-body', styles.content)}>
                  {message && message.body}
                </div>
              </article>
            </CSSTransition>
          ))}
        </TransitionGroup>
      </div>
    );
  }
}
