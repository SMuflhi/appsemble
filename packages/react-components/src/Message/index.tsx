import type { BulmaColor } from '@appsemble/sdk';
import classNames from 'classnames';
import React, { ReactElement, ReactNode } from 'react';

interface MessageProps {
  /**
   * The message content.
   */
  children: ReactNode;

  /**
   * Additional class names to assign to the message element.
   */
  className?: string;

  /**
   * The message type.
   */
  color?: BulmaColor;

  /**
   * An optional header for the message.
   */
  header?: ReactNode;
}

export default function Message({
  children,
  className,
  color,
  header,
}: MessageProps): ReactElement {
  return (
    <div className={classNames('message', className, { [`is-${color}`]: color })}>
      {header && <h6 className="message-header">{header}</h6>}
      <div className="message-body">{children}</div>
    </div>
  );
}
