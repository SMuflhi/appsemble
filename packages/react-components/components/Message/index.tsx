import { injectIntl } from 'react-intl';

import Message, { MessageProps } from './Message';

export default injectIntl<'intl', MessageProps>(Message);
