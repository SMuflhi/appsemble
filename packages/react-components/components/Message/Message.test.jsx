import React from 'react';
import { shallow } from 'enzyme';

import Message from './Message';

describe('Message', () => {
  it('should render messages', () => {
    const messages = [{ id: 1, body: 'Foo' }, { id: 2, body: 'Bar', color: 'info' }];
    const result = shallow(<Message messages={messages} remove={() => {}} />);

    expect(result).toMatchSnapshot();
  });

  it('should call remove after 5 seconds', () => {
    jest.useFakeTimers();

    const messages = [{ id: 1, body: 'Foo' }, { id: 2, body: 'Bar', color: 'info' }];
    const mock = jest.fn(() => {});
    const result = shallow(<Message messages={[]} remove={mock} />);

    // Simulate adding messages over time
    result.setProps({ messages: [messages[0]] });
    jest.advanceTimersByTime(1000);
    result.setProps({ messages });

    expect(mock.mock.calls).toHaveLength(0);

    jest.advanceTimersByTime(4000);
    expect(mock.mock.calls).toHaveLength(1);
    expect(mock.mock.calls[0][0]).toStrictEqual(messages[0]);

    jest.advanceTimersByTime(1000);
    expect(mock.mock.calls).toHaveLength(2);
    expect(mock.mock.calls[1][0]).toStrictEqual(messages[1]);
  });
});
