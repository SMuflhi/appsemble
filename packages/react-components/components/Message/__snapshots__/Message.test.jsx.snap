// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render messages 1`] = `
<div
  className="root"
>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "messageEnter",
          "enterActive": "messageEnterActive",
          "exit": "messageExit",
          "exitActive": "messageExitActive",
        }
      }
      key="1"
      timeout={300}
    >
      <article
        className="message is-danger"
      >
        <div
          className="message-body content"
        >
          Foo
        </div>
      </article>
    </CSSTransition>
    <CSSTransition
      classNames={
        Object {
          "enter": "messageEnter",
          "enterActive": "messageEnterActive",
          "exit": "messageExit",
          "exitActive": "messageExitActive",
        }
      }
      key="2"
      timeout={300}
    >
      <article
        className="message is-info"
      >
        <div
          className="message-body content"
        >
          Bar
        </div>
      </article>
    </CSSTransition>
  </TransitionGroup>
</div>
`;
